File: problem0035.py
# The number, 197, is called a circular prime because all rotations of the
# digits: 197, 971, and 719, are themselves prime.
# 
# There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37,
# 71, 73, 79, and 97.
# 
# How many circular primes are there below one million?

import numbertheory as nt
import time

def extract(n):
    return [d for d in str(n)]

def rotate(it):
    L = len(it)
    rotit = [it[(i+1)%L] for i in range(L)]
    return rotit

def recombine(digits):
    return int(''.join(digits))

def isCircular(n):
    if not nt.isPrime(n):
        return False
    nDigits = extract(n)
    pDigits = rotate(nDigits)
    p = recombine(pDigits)
    while p != n:
        if not nt.isPrime(p):
            return False
        pDigits = rotate(pDigits)
        p = recombine(pDigits)
    return True

start_time = time.clock()

upperBound = 1000000
primes = nt.allPrimesLessThan(upperBound)

checkpoint1 = time.clock()

print 'all primes less than {0} computed'.format(upperBound)
print '{0} seconds elapsed'.format(checkpoint1-start_time)

tot = 0
for p in primes:
    if isCircular(p):
        tot += 1
print 'number of circular primes: {0}'.format(tot)

end_time = time.clock()

print '{0} seconds elapsed'.format(end_time-start_time)

File: problem0035.java
/*
The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.

There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.

How many circular primes are there below one million?
*/

import java.util.ArrayList;
import java.util.List;

public class problem0035 {
	public static List<Integer> primes = null;
	
	public static void main(String[] args) {
		final int UPPER_BOUND = 1000000;
		primes = allPrimesLessThan(UPPER_BOUND);
		System.out.println(primes.size());
	}
	
	public static List<Integer> divisors(int n) {
		List<Integer> divs = new ArrayList<Integer>();
		for( int i = 1; i <= Math.sqrt(n); i++ ) {
			if( n % i == 0 ) {
				divs.add(divs.size(),i);
				if( n/i != i ) divs.add(divs.size(),n/i);
			}
		}
		divs.sort(null); // sorts list of divisors in ascending order
		return divs;
	}
	
	public static int sumOfDivisors(int n) {
		List<Integer> divs = divisors(n);
		int sum = 0;
		for( int i = 0; i < divs.size(); i++ ) {
			sum += divs.get(i);
		}
		return sum;
	}
	
	public static int sumOfProperDivisors(int n) {
		return sumOfDivisors(n) - n;
	}
	
	public static boolean isPrime(int n) {
		return sumOfProperDivisors(n) == 1;
	}
	
	public static List<Integer> allPrimesLessThan(int n) {
		List<Integer> primes = new ArrayList<Integer>();
		for( int i = 2; i < n; i++ ) {
			if( isPrime(i) ) {
				primes.add(i);
			}
		}
		return primes;
	}
	
}